### ISSUE TITLE
Occasional crash in parrot integration when switching microphones to None

### ISSUE THREAD

**Issue by chaosparrot** [talonvoice/talon#376]
While developing a one click icon to toggle my current mic from last used to None, I came across an occasional crash when it would set the microphone to None. This would eventually lead to a Talon crash a few seconds later.

I haven't been able to find any reliable patterns when it crashes. 

Talon version: 0.2.0-11-g8007
System: Windows 10
Microphones available at that time, Microphone (HD Pro Webcam C920) as System Default. HyperX Quadcast as other Microphone.

IO log contains log messages with currently active microphone, and the microphone to transition to.

```
2021-08-17 21:30:46  INFO Activating Microphone: "Microphone (HD Pro Webcam C920)"
2021-08-17 21:30:47    IO Microphone (HD Pro Webcam C920) Microphone (HD Pro Webcam C920)
2021-08-17 21:30:47  INFO Deactivating Microphone
2021-08-17 21:30:47    IO None Microphone (HD Pro Webcam C920)
2021-08-17 21:30:47  INFO Activating Microphone: "Microphone (HD Pro Webcam C920)"
2021-08-17 21:30:47    IO Microphone (HD Pro Webcam C920) Microphone (HD Pro Webcam C920)
2021-08-17 21:30:47  INFO Deactivating Microphone
2021-08-17 21:30:48    IO None Microphone (HD Pro Webcam C920)
2021-08-17 21:30:48  INFO Activating Microphone: "Microphone (HD Pro Webcam C920)"
2021-08-17 21:30:48    IO Microphone (HD Pro Webcam C920) Microphone (HD Pro Webcam C920)
2021-08-17 21:30:48  INFO Deactivating Microphone
2021-08-17 21:30:48    IO None Microphone (HD Pro Webcam C920)
2021-08-17 21:30:48  INFO Activating Microphone: "Microphone (HD Pro Webcam C920)"
2021-08-17 21:30:49    IO Microphone (HD Pro Webcam C920) Microphone (HD Pro Webcam C920)
2021-08-17 21:30:49  INFO Deactivating Microphone
2021-08-17 21:30:49    IO None Microphone (HD Pro Webcam C920)
2021-08-17 21:30:49  INFO Activating Microphone: "Microphone (HD Pro Webcam C920)"
2021-08-17 21:30:49    IO Microphone (HD Pro Webcam C920) Microphone (HD Pro Webcam C920)
2021-08-17 21:30:50  INFO Deactivating Microphone
2021-08-17 21:30:50    IO None Microphone (HD Pro Webcam C920)
2021-08-17 21:30:50  INFO Activating Microphone: "Microphone (HD Pro Webcam C920)"
2021-08-17 21:30:50    IO Microphone (HD Pro Webcam C920) Microphone (HD Pro Webcam C920)
2021-08-17 21:30:50  INFO Deactivating Microphone
2021-08-17 21:30:50    IO None Microphone (HD Pro Webcam C920)
2021-08-17 21:30:50  INFO Activating Microphone: "Microphone (HD Pro Webcam C920)"
2021-08-17 21:30:51    IO Microphone (HD Pro Webcam C920) Microphone (HD Pro Webcam C920)
2021-08-17 21:30:51  INFO Deactivating Microphone
2021-08-17 21:30:51    IO None Microphone (HD Pro Webcam C920)
2021-08-17 21:30:51  INFO Activating Microphone: "Microphone (HD Pro Webcam C920)"
2021-08-17 21:30:51    IO Microphone (HD Pro Webcam C920) Microphone (HD Pro Webcam C920)
2021-08-17 21:30:51  INFO Deactivating Microphone
2021-08-17 21:30:51    IO None Microphone (HD Pro Webcam C920)
2021-08-17 21:30:51  INFO Activating Microphone: "Microphone (HD Pro Webcam C920)"
2021-08-17 21:30:52    IO Microphone (HD Pro Webcam C920) Microphone (HD Pro Webcam C920)
2021-08-17 21:30:52  INFO Deactivating Microphone
2021-08-17 21:30:52    IO None Microphone (HD Pro Webcam C920)
2021-08-17 21:30:52  INFO Activating Microphone: "Microphone (HD Pro Webcam C920)"
2021-08-17 21:30:52    IO Microphone (HD Pro Webcam C920) Microphone (HD Pro Webcam C920)
2021-08-17 21:30:52  INFO Deactivating Microphone
2021-08-17 21:30:52    IO None Microphone (HD Pro Webcam C920)
2021-08-17 21:30:52  INFO Activating Microphone: "Microphone (HD Pro Webcam C920)"
2021-08-17 21:30:53    IO Microphone (HD Pro Webcam C920) Microphone (HD Pro Webcam C920)
2021-08-17 21:30:53  INFO Deactivating Microphone
2021-08-17 21:30:53    IO None Microphone (HD Pro Webcam C920)
2021-08-17 21:30:54  INFO Activating Microphone: "Microphone (HD Pro Webcam C920)"
2021-08-17 21:30:54    IO Microphone (HD Pro Webcam C920) Microphone (HD Pro Webcam C920)
2021-08-17 21:30:54  INFO Deactivating Microphone
2021-08-17 21:30:54    IO None Microphone (HD Pro Webcam C920)
2021-08-17 21:30:54  INFO Activating Microphone: "Microphone (HD Pro Webcam C920)"
2021-08-17 21:30:54    IO Microphone (HD Pro Webcam C920) Microphone (HD Pro Webcam C920)
2021-08-17 21:30:54  INFO Deactivating Microphone
2021-08-17 21:30:54    IO None Microphone (HD Pro Webcam C920)
2021-08-17 21:30:55  INFO Activating Microphone: "Microphone (HD Pro Webcam C920)"
2021-08-17 21:30:57    IO Microphone (HD Pro Webcam C920) Microphone (HD Pro Webcam C920)
2021-08-17 21:30:57  INFO Deactivating Microphone
2021-08-17 21:30:59    IO None Microphone (HD Pro Webcam C920)
2021-08-17 21:30:59  INFO Activating Microphone: "Microphone (HD Pro Webcam C920)"
2021-08-17 21:31:00    IO Microphone (HD Pro Webcam C920) Microphone (HD Pro Webcam C920)
2021-08-17 21:31:00  INFO Deactivating Microphone
2021-08-17 21:31:01    IO None Microphone (HD Pro Webcam C920)
2021-08-17 21:31:01  INFO Activating Microphone: "Microphone (HD Pro Webcam C920)"
2021-08-17 21:31:03    IO Microphone (HD Pro Webcam C920) Microphone (HD Pro Webcam C920)
2021-08-17 21:31:03  INFO Deactivating Microphone
2021-08-17 21:31:05 WARNING [watchdog] "user.talon_hud.widgets.statusbar.on_mouse" @2.0s
   26:                        threading.py:912* # cron thread
   25:                        threading.py:954* 
   24:                        threading.py:892* 
   23:                       talon\cron.py:155| 
   22:                       talon\cron.py:106| 
   21:         talon\scripting\dispatch.py:99 | 
   20:         talon\scripting\dispatch.py:105| 
   19:         talon\scripting\dispatch.py:144| 
   18:         talon\scripting\dispatch.py:135| 
   17:             talon\scripting\rctx.py:233| # 'mouse' user.talon_hud.display:on_mouse()
   16: user\talon_hud\widgets\statusbar.py:123| self.activate_icon()
   15: user\talon_hud\widgets\statusbar.py:133| actions.user.activate_statusbar_icon(s..
   14:          talon\scripting\actions.py:75 | # user.activate_statusbar_icon()
   13:            talon\scripting\types.py:366| 
   12: user\talon_hud\widgets\statusbar.py:307| actions.sound.set_microphone("None")
   11:          talon\scripting\actions.py:75 | # sound.set_microphone()
   10:            talon\scripting\types.py:366| 
    9:       talon\scripting\core\sound.py:31 | 
    8:         talon\plugins\microphone.py:99 | 
    7:         talon\plugins\microphone.py:153| 
    6:         talon\scripting\dispatch.py:105| 
    5:         talon\scripting\dispatch.py:144| 
    4:         talon\scripting\dispatch.py:135| 
    3:             talon\scripting\rctx.py:233| # 'mic_change' user.parrot_integration:_set_microphone()
    2: talon\experimental\parrot\system.py:182| 
    1:                  talon\lib\cubeb.py:530| 
2021-08-17 21:31:05 WARNING [watchdog] "talon.experimental.parrot.system._set_microphone" @2.0s
   26:                        threading.py:912* # cron thread
   25:                        threading.py:954* 
   24:                        threading.py:892* 
   23:                       talon\cron.py:155| 
   22:                       talon\cron.py:106| 
   21:         talon\scripting\dispatch.py:99 | 
   20:         talon\scripting\dispatch.py:105| 
   19:         talon\scripting\dispatch.py:144| 
   18:         talon\scripting\dispatch.py:135| 
   17:             talon\scripting\rctx.py:233| # 'mouse' user.talon_hud.display:on_mouse()
   16: user\talon_hud\widgets\statusbar.py:123| self.activate_icon()
   15: user\talon_hud\widgets\statusbar.py:133| actions.user.activate_statusbar_icon(s..
   14:          talon\scripting\actions.py:75 | # user.activate_statusbar_icon()
   13:            talon\scripting\types.py:366| 
   12: user\talon_hud\widgets\statusbar.py:307| actions.sound.set_microphone("None")
   11:          talon\scripting\actions.py:75 | # sound.set_microphone()
   10:            talon\scripting\types.py:366| 
    9:       talon\scripting\core\sound.py:31 | 
    8:         talon\plugins\microphone.py:99 | 
    7:         talon\plugins\microphone.py:153| 
    6:         talon\scripting\dispatch.py:105| 
    5:         talon\scripting\dispatch.py:144| 
    4:         talon\scripting\dispatch.py:135| 
    3:             talon\scripting\rctx.py:233| # 'mic_change' user.parrot_integration:_set_microphone()
    2: talon\experimental\parrot\system.py:182| 
    1:                  talon\lib\cubeb.py:530| 
```

**Comment by lunixbochs** [comment]
Can you go to your system event log and find the crashing module and offset, and post it along with the talon version in question?

**Comment by chaosparrot** [comment]
This is the error I find in the Event Viewer (Talon version: 0.2.0-11-g8007):

```Faulting application name: talon_console.exe, version: 1.0.0.0, time stamp: 0x00000000
Faulting module name: ntdll.dll, version: 10.0.19041.1110, time stamp: 0xe7a22463
Exception code: 0xc0000005
Fault offset: 0x0000000000063416
Faulting process ID: 0x2f24
Faulting application start time: 0x01d7939e2fd2db2b
Faulting application path: C:\Program Files\Talon\talon_console.exe
Faulting module path: C:\WINDOWS\SYSTEM32\ntdll.dll
Report ID: 9be95250-4fee-4ea5-ac16-8ca218a5d96c
Faulting package full name: 
Faulting package-relative application ID: 
```

Info messages that happens right after:

```
Fault bucket , type 0
Event Name: APPCRASH
Response: Not available
Cab Id: 0

Problem signature:
P1: talon_console.exe
P2: 1.0.0.0
P3: 00000000
P4: ntdll.dll
P5: 10.0.19041.1110
P6: e7a22463
P7: c0000005
P8: 0000000000063416
P9: 
P10: 
```

**Comment by chaosparrot** [comment]
I cannot seem to reproduce this bug in the latest beta. I have tried to generate a crash by agressively whistling, changing modes, changing mics etc for about five minutes. I used to be able to reliably reproduce this in just about a minute of toggling microphones.

As far as I am concerned this issue can be closed. If I do happen to encounter a similar crash again, I will make a note for myself to reopen this ticket.
